### ISSUE TITLE
Content on dynamic pages reverts to id on built versions

### ISSUE THREAD

**Issue by romeboards** [mccrodp/gatsby-starter-contentful-i18n#6]
Noticing a strange error exclusively on built/hosted versions that seems to be related to earlier issues with dynamic pages from Contentful - 

To reproduce:
1. Navigate directly to the page in question, i.e. 
https://gatsby-starter-contentful-i18n.netlify.com/de/products/c5ksdbwsexy6qegucyaoacs/

2. The necessary content ("Wählen Sie Ihre Sprache") will display briefly before reverting to "selectLanguage"

The messages are getting appropriately sent to IntlProvider, but children components that have injectIntl wrappers don't seem to be getting the appropriately set locale.

**Comment by mccrodp** [comment]
Ohh, well spotted. Interestingly, it's working on the `en-US` route. https://gatsby-starter-contentful-i18n.netlify.com/en-US/products/c5ksdbwsexy6qegucyaoacs/

I suggest taking a look at the hack here and see why we aren't getting the correct data maintained for `de`, but, as you say, it's reverting, so I'm not sure yet. 
https://github.com/mccrodp/gatsby-starter-contentful-i18n/blob/master/src/layouts/index.js#L22

Feel free to submit a PR if you find out in the meantime. Cheers!

**Comment by romeboards** [comment]
Given that the issue only occurs in the built/compiled version of the app, I'm curious if moving the language data to the [static folder](https://www.gatsbyjs.org/docs/adding-images-fonts-files/#using-the-static-folder) might make sense. Although its strange that it would only occur on dynamic pages as well if that were the case.

**Comment by mccrodp** [comment]
Ahh, now I get you, only happening in build, missed that. No wonder I didn't spot it. I'm not sure yet, but worth a try? This whole lang selector on dynamic pages is still a hack, would love if the upstream issue could be found, but alas, I've limited time on this at the moment. Anything you find much appreciated :)

**Comment by mccrodp** [comment]
What's interesting about this issue also is if you navigate to the home page, change the locale to `de` and navigate to a product page, it works, but if you click Refresh, it doesn't have access to the string. It's some kind of caching issue. That perhaps means at the point where the hack is, the messages string has not been loaded into memory, or not accessible somehow, but it is when in development mode.

**Comment by romeboards** [comment]
Yeah the children components (wrapped in injectIntl) are failing to recognize the set locale:

![image](https://user-images.githubusercontent.com/627751/37316210-325e44f0-2634-11e8-99b1-22e7fca03791.png)

What's interesting is that they're reverting to "en" as opposed to "en-US" which is the default. 

Moving the message files to a static folder didn't work. I'll keep investigating as much as I can...

**Comment by mccrodp** [comment]
Huhh, that is interesting, sounds like you're getting somewhere though. I wonder what `getCurrentLang` is returning  https://github.com/mccrodp/gatsby-starter-contentful-i18n/blob/master/src/layouts/index.js#L17

Also, I believe the issue may really stem from the fact that in TemplateWrapper is not passed the correct props (incl. `i18nMessages` where our issue likely lies) when being used by the dynamic pages: `product` and `category`, hence the hack. Any ideas on that might reach a fix and an improvement on current hack https://github.com/mccrodp/gatsby-starter-contentful-i18n/blob/master/src/layouts/index.js#L14

**Comment by mccrodp** [comment]
Hey, @romeboards, how are you debugging Gatsby? I'm not 100% how to best to debug at the moment, would love to have a debugger setup rather than using `console.log`s. What are you dumping there when you refer to `injectIntl` in the image there and where in code? I cannot see that at the moment.

I think the comments from the below link onwards may give some good tips on how we can remodel things in the layouts, perhaps using a layout component and then push those changes upstream to the i18n community plugin, otherwise the method from this repo in this comment is worth exploring https://github.com/gatsbyjs/gatsby/issues/3853#issuecomment-367038918

**Comment by mccrodp** [comment]
My lack of understanding at the moment is for the dynamic content: product and category. How is the `TemplateWrapper` in Layout invoked? The i18n plugin usually routes through the lang specific layout for certain pages, but not dynamic pages, right? That's why we get undefined for that language specific message, then need the hack. If we knew where that was invoked, perhaps you can pass the i18nMessages from there, but I'm not sure. End of my day here and the small amount of brain cells left wasn't able to do much with it. Might be a good idea to create an issue upstream for how to ensure that dynamic pages point to the language specific layouts and don't call `index.js` directly?

**Comment by romeboards** [comment]
@mccrodp The printout is from the [react devtools](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) extension on Chrome! very useful.

I think your second point is right on - I'm currently messing around with the `useLangKeyLayout` and `pagesPaths` options on gatsby-plugin-i18n as well as looking a little deeper into their codebase. I'll be sure to keep you updated if I find anything useful for this issue...

**Comment by romeboards** [comment]
Looks like I found a temporary fix thats a slightly altered version of the current hack:

In `gatsby-config.js`:

Use `useLangKeyLayout: false` to disable language specific layouts and just rely on `index.js`

Then in `index.js`:

Incorporate all of the locale data at once:
```
import en from 'react-intl/locale-data/en';
import es from 'react-intl/locale-data/es';

addLocaleData([...en, ...es]);
```

Then just rely exclusively on the conditional you put in place for the dynamic pages:

```
  let i18nMessages;
  if (langKey === 'es') {
    i18nMessages = messagesEs
  } else {
    i18nMessages = messagesEn
  }
```

WIll keep looking for a more sustainable solution, but this seems to resolve the current issue at hand...

**Comment by mccrodp** [comment]
Nice one, can you submit a PR and let's get this in the codebase for now to get build working properly.

**Comment by romeboards** [comment]
done - take a look on your end and make sure everything checks out!

In the future we can probably create a file to instantiate the locale data outside of index.js, but this seems clean enough for now.

**Comment by mccrodp** [comment]
Fixed via #7, thanks!
